C/services/filter-plugin-interfaces/python/filter_ingest_pymodule/ingest_callback_pymodule.cpp

```diff
@@ -0,0 +1,158 @@
+
+/**
+ * Implementation of data ingest into filters chain
+ *
+ * @param    self       The python module object
+ * @param    args       Input arguments
+ * @return              PyObject of None type
+ */
+static PyObject *filter_ingest_callback(PyObject *self, PyObject *args)
+{
+	PyObject *readingList;
+	PyObject *callback;
+	PyObject *ingestData;
+
+	if (!PyArg_ParseTuple(args,
+			      "OOO",
+			      &callback,
+			      &ingestData,
+			      &readingList))
+	{
+		return NULL;
+	}
+
+	// Invoke callback routine
+	filter_plugin_ingest_fn(callback,
+				ingestData,
+				readingList);
+
+	Py_INCREF(Py_None);
+	return Py_None;
+}
+
+static PyMethodDef FilterIngestMethods[] = {
+	{
+		"filter_ingest_callback",
+		filter_ingest_callback,
+		METH_VARARGS,
+		"Invoke filter ingest callback"
+	},
+	{NULL, NULL, 0, NULL}    /* Sentinel */
+};
+
+static struct PyModuleDef filterIngestmodule = {
+	PyModuleDef_HEAD_INIT,
+	"filter_ingest",   /* name of module */
+	NULL, 		/* module documentation, may be NULL */
+	-1,       	/* size of per-interpreter state of the module,
+	             or -1 if the module keeps state in global variables. */
+	FilterIngestMethods
+};
+
+/**
+ * Init the C API Python module
+ */
+PyMODINIT_FUNC
+PyInit_filter_ingest(void)
+{	
+	PyObject *m;
+
+	m = PyModule_Create(&filterIngestmodule);
+	if (m == NULL)
+		return NULL;
+
+	IngestError = PyErr_NewException("ingest.error", NULL, NULL);
+	Py_INCREF(IngestError);
+	PyModule_AddObject(m, "error", IngestError);
+
+	return m;
+}
+
+/**
+ * Ingest data into filters chain
+ *
+ * @param    ingest_callback            The callback routine
+ * @param    ingest_obj_ref_data        Object parameter for callback routine
+ * @param    readingsObj                Readongs data as PyObject
+ */
+void filter_plugin_ingest_fn(PyObject *ingest_callback,
+			     PyObject *ingest_obj_ref_data,
+			     PyObject *readingsObj)
+{
+	PRINT_FUNC;
+	if (ingest_callback == NULL ||
+	    ingest_obj_ref_data == NULL ||
+	    readingsObj == NULL)
+	{
+		Logger::getLogger()->error("PyC interface error: "
+					   "filter_plugin_ingest_fn: "
+					   "filter_ingest_callback=%p, "
+					   "ingest_obj_ref_data=%p, "
+					   "readingsObj=%p",
+					   ingest_callback,
+					   ingest_obj_ref_data,
+					   readingsObj);
+		return;
+	}
+
+	std::vector<Reading *> *vec = NULL;
+
+	// Check we have a list of readings
+	if (PyList_Check(readingsObj))
+	{
+		// Get vector of Readings from Python object
+		vec =  Py2C_getReadings(readingsObj);
+	}
+	else
+	{
+		Logger::getLogger()->error("Filter did not return a Python List "
+					   "but object type %s",
+					   Py_TYPE(readingsObj)->tp_name);
+	}
+
+	if (vec)
+	{
+		// Get callback pointer
+		INGEST_CB_DATA cb = (INGEST_CB_DATA) PyCapsule_GetPointer(ingest_callback, NULL);
+		// Get ingest object parameter
+		void *data = PyCapsule_GetPointer(ingest_obj_ref_data, NULL);
+
+		// Create ReadingSet object
+		ReadingSet *newData = new ReadingSet(vec);
+
```

Comment:
I think 'vec' can be freed here.

Reply:
It's added to ReadingSet at it will be deleted by ReadingSet destructor

Reply:
I am talking about deleting vec, not readings inside vector/ReadingSet. This vector has been allocated here:

std::vector<Reading *>* Py2C_getReadings(PyObject *polledData)
{
	std::vector<Reading *>* newReadings = new std::vector<Reading *>();